### PR TITLE
Handle RecursionError better and centralize protocol error handling

### LINKED ISSUE #4677: Creating deep nested type gives error `could not serialize result in worker subprocess`
- EdgeDB Version:
EdgeDB CLI 2.2.0+35980b5
EdgeDB Instance    "version": "2.6+db69671",
- OS Version:
MacOS 12.6 (21G115)

The schema has some types connected by links.
For each link to type `Target` the default is a freshly created as `(insert Target)`.
Most likely, the system can only cope with very short chains of such inserts or the total number of the inserts is too high.

Steps to Reproduce:

1. `edgedb migration create --non-interactive` (takes 2-3 minutes) 
2. `edgedb migrate` (takes a bit more than 2 minutes)
3. echo 'insert `varycon.Henkel.Adhesive Experience`::`ARMasterConfig` { id := <uuid>"dc29dd8b-6a44-47de-9048-a2d58629a193" };' | edgedb

Result:

edgedb error: InternalServerError: could not serialize result in worker subprocess
  Hint: This is most likely a bug in EdgeDB. Please consider opening an issue ticket at https://github.com/edgedb/edgedb/issues/new?template=bug_report.md
  Server traceback:
      Can't pickle local object 'process_insert_body.<locals>.dynamic_get_path':

      Traceback (most recent call last):
        File "/Users/doc/Library/Application Support/edgedb/portable/2.6/lib/python3.10/site-packages/edb/server/compiler_pool/worker_proc.py", line 64, in worker
          pickled = pickle.dumps(data, -1)
      AttributeError: Can't pickle local object 'process_insert_body.<locals>.dynamic_get_path'

Schema:

[default.esdl.zip](https://github.com/edgedb/edgedb/files/10021869/default.esdl.zip)

### PR DESCRIPTION
**Pull request by msullivan**
Make stack overflows consistently produce an UnsupportedFeatureError
instead of producing various ISEs.
They aren't the same class of compiler/server bugs as other ISEs,
typically.

In order to handle this properly for *all* our protocols, add a
centralized interpret_error function to execute. It turns out that
http and notebook weren't interpreting backend errors at all.

Closes #4677.

### CODE REVIEW

**Comment by msullivan** [comment]
Going to have to rework this a decent amount now that #5921 is in, but this version will more or less go into 3.x

**Comment by elprans** [comment]
This needs a rebase

File: edb/server/dbview/dbview.pyx [hunk 1]
```diff
@@ -1074,14 +1074,14 @@ cdef class DatabaseConnectionView:
...
 
-    async def interpret_backend_error(self, exc):
-        compiler_pool = self._db._index._server.get_compiler_pool()
-        return await compiler_pool.interpret_backend_error(
-            self.get_user_schema_pickle(),
-            self.get_global_schema_pickle(),
-            exc.fields,
-            False,
-        )
+    # async def interpret_backend_error(self, exc):
```
Comment (elprans) [inline edb/server/dbview/dbview.pyx:1077]: ?

File: edb/server/protocol/execute.pyx [hunk 2]
```diff
@@ -585,3 +586,64 @@ cdef bytes _encode_args(list args):
...
+
+            exc = RuntimeError(
+                'unhandled error while calling interpret_backend_error(); '
+                'run with EDGEDB_DEBUG_SERVER to debug.')
+
+    exc_type = type(exc)
+    if not issubclass(exc_type, errors.EdgeDBError):
+        exc_type = errors.InternalServerError
+
+    return exc, exc_type
```
Comment (elprans) [inline edb/server/protocol/execute.pyx:649]: Can you elaborate a bit please why this instead of chaining or wrapping?

Comment (msullivan) [reply edb/server/protocol/execute.pyx:649]: Let me try changing it

Comment (msullivan) [reply edb/server/protocol/execute.pyx]: Changed

**Review by elprans — APPROVED** [review]

**Pull request merged**